### ISSUE TITLE
To use multiple connectionID's/CIP instances using ForwardOpen

### ISSUE THREAD

**Issue by rakesh1225** [EIPStackGroup/OpENer#369]
Hello

We are using Opener version 2.1.0 on our Linux device. We are able to use it properly for sending/receiving single frame of assembly data with single connection ID each between 2 devices(Scanner1 and Opener adapter1) using CIP.

We have a requirement to send 3 frames of different assembly data(like assembly1,assembly2,assembly3) with different connection ID's/CIP instance for assembly data between 2 devices(Scanner1 and Opener adapter1) using CIP.

Please let me know if we have to increase the number of ForwardOpens for multiple assemblies with unique cip instance each? 
What changes may be required in configuring the Opener stack in this case?
Do I need to add more assembly messages from Openerstack like to increase the below code:

   // create 3 assembly object instances
    auto *assembly_data064 = cip_app.GetAssemblyData064();
    auto *assembly_data096 = cip_app.GetAssemblyData096();
    auto *assembly_data097 = cip_app.GetAssemblyData097();
    auto *assembly_data09A = cip_app.GetAssemblyData09A();

    // INPUT
    CreateAssemblyObject(DEMO_APP_INPUT_ASSEMBLY_NUM,
                         assembly_data064,
                         INPUT_ASSEMBLY_SIZE);

    // OUTPUT
    CreateAssemblyObject(DEMO_APP_OUTPUT_ASSEMBLY_NUM,
                         assembly_data096,
                         OUTPUT_ASSEMBLY_SIZE);

    // CONFIG
    CreateAssemblyObject( DEMO_APP_CONFIG_ASSEMBLY_NUM,
                         assembly_data097,
                         sizeof(assembly_data097));

**Comment by CapXilinx** [comment]
Hi,
I don't understand your requirement. Could you please try to provide an example/figure of it.

Connections IDs are not chosen by the adapter, but by the scanner afaik

Best regards,
Martin

**Comment by CapXilinx** [comment]
@rakesh1225 OpENer also supports the LargeForwardOpen, which allows larger payloads.

If LargeForwardOpen is not an option, you can add additional assemblies and create Input-only connections to them (by selecting the Input Only assembly as O->T target)

**Comment by rakesh1225** [comment]
Yes, planning to use ForwardOpen only.

Does Input-only connections allow mulitple assemblies? As per the following function definition it allows only 0 as connection number and stores in array. It is discarding anything more than that.
ConfigureInputOnlyConnectionPoint(0,
                                      DEMO_APP_HEARTBEAT_INPUT_ONLY_ASSEMBLY_NUM,
                                      DEMO_APP_INPUT_ASSEMBLY_NUM1,
                                      DEMO_APP_CONFIG_ASSEMBLY_NUM);
ConfigureInputOnlyConnectionPoint(1,
                                      DEMO_APP_HEARTBEAT_INPUT_ONLY_ASSEMBLY_NUM,
                                      DEMO_APP_INPUT_ASSEMBLY_NUM2,
                                      DEMO_APP_CONFIG_ASSEMBLY_NUM);
ConfigureInputOnlyConnectionPoint(2,
                                      DEMO_APP_HEARTBEAT_INPUT_ONLY_ASSEMBLY_NUM,
                                      DEMO_APP_INPUT_ASSEMBLY_NUM3,
                                      DEMO_APP_CONFIG_ASSEMBLY_NUM);

![image](https://user-images.githubusercontent.com/3583648/162475732-0318643d-95f6-42ed-81c4-3bf9d64e6580.png)

**Comment by CapXilinx** [comment]
No, you can only target one assembly with each connection. There is no way around that.
You would need 3 Input Assemblies and 1 output assemblies in your case, which means you will create one Exclusive Owner connections, and two additional Input-Only connection to get all of your data.

For that you have to tune the defines in the user conf

**Comment by azoitl** [comment]
The CIP protocol does not really allow what you would like to have. If you have three assemblies to be sent from your device to the scanner you need to have three forward opens. However what you could have is that you have one normal input/output connection in your example 101 & 100 would be as @CapXilinx  just wrote be the exclusive owner connection and you could have two additional input only connections. However for the input only connection you also need recieving assemblies. These receiving assemblies can AFAIR be of lenght 0 but CIP need it to identify the connection. The receiving assemblies are used for the heart beat messages which have to be sent from the scanner to your devices. This is used by OpENer to detect timeouts like when the scanner is gone or cables are broken. I hope this helps.

**Comment by rakesh1225** [comment]
Hello @CapXilinx , @azoitl 
Thanks for the response.

I will do 3 ForwardOpen's from Scanner.
So the following changes will be done from Adapter/Opener. However I don't find a way to send DEMO_APP_HEARTBEAT_INPUT_ONLY_ASSEMBLY_NUM. 

Will there be any issue to connection if HEARTBEAT will not be sent from Adapter to Scanner? or any other way to send HEARTBEAT from Adapter to scanner.

   ConfigureExclusiveOwnerConnectionPoint(0,
                                           DEMO_APP_OUTPUT_ASSEMBLY_NUM,
                                           DEMO_APP_INPUT_ASSEMBLY_NUM_1,
                                           DEMO_APP_CONFIG_ASSEMBLY_NUM);

    ConfigureInputOnlyConnectionPoint(0,
                                      DEMO_APP_HEARTBEAT_INPUT_ONLY_ASSEMBLY_NUM,
                                      DEMO_APP_INPUT_ASSEMBLY_NUM_2,
                                      DEMO_APP_CONFIG_ASSEMBLY_NUM);

    ConfigureInputOnlyConnectionPoint(1,
                                      DEMO_APP_HEARTBEAT_INPUT_ONLY_ASSEMBLY_NUM,
                                      DEMO_APP_INPUT_ASSEMBLY_NUM_3,
                                      DEMO_APP_CONFIG_ASSEMBLY_NUM);

**Comment by CapXilinx** [comment]
The heartbeat sends no payload data. It only ensures that the other side is still alive

**Comment by azoitl** [comment]
In your case the heartbeat is sent from the scanner to the adapter. In the other direction you have the input only data. However you rconfiguration will not work in the intended way. You need at least 2 different Heartbeat input only assemblies.

**Comment by rakesh1225** [comment]
Thank you so much for the support. @azoitl, @CapXilinx 
Since the assembly data size is keep increasing as per our requirement, we are trying to use large forward open instead of multiple forward open's.

Large forward open is working good with assembly data  size less than 1444 only. But when I tried to increase the large forward open size to 1603 bytes, it is failing to send data.

Note: I increased the macro PC_OPENER_ETHERNET_BUFFER_SIZE size to 1603 already. Still couldn't send data size of 1603 over CIP using large forward open.

Pl confirm if any more changes required to increase large forward open size in opener stack.
